File: pipeline/shared/src/commonMain/kotlin/com/informedcitizen/pipeline/model/RollCallVote.kt
package com.informedcitizen.pipeline.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** Chamber a roll-call vote was taken in. */
@Serializable
enum class Chamber {
    @SerialName("house") HOUSE,
    @SerialName("senate") SENATE,
}

/**
 * A member's recorded position on a roll call. Upstream sources use
 * varying labels (House "Aye"/"No" on motions, "Yea"/"Nay" on
 * passage; Senate "Guilty"/"Not Guilty" on impeachment); the pipeline
 * normalizes all of them to these four wire values.
 */
@Serializable
enum class VotePosition {
    @SerialName("yea") YEA,
    @SerialName("nay") NAY,
    @SerialName("present") PRESENT,
    @SerialName("not_voting") NOT_VOTING,
}

/** Chamber-wide tallies for one roll call. */
@Serializable
data class VoteTotals(
    val yea: Int,
    val nay: Int,
    val present: Int,
    @SerialName("not_voting") val notVoting: Int,
)

/** One member's position on one roll call, keyed by bioguide id. */
@Serializable
data class MemberVote(
    @SerialName("bioguide_id") val bioguideId: String,
    val position: VotePosition,
    val party: String? = null,
    val state: String? = null,
)

/**
 * One roll-call vote with per-member positions, published at
 * `docs/data/votes/congress<N>/<chamber>-<session>-<rollNumber>.json`
 * (e.g. `votes/congress119/house-1-17.json`).
 *
 * Votes are sharded one-file-per-vote rather than bundled per
 * Congress: positions run ~435 rows for a House vote and a Congress
 * holds 1,000+ roll calls, so a bundled file would be tens of MB
 * while the app only ever needs the handful of votes attached to the
 * bill on screen. The per-Congress [VotesIndex] carries everything
 * except positions for list rendering and bill lookup.
 *
 * [id] is `<chamber>-<congress>-<session>-<rollNumber>`
 * (e.g. `house-119-1-17`). [billId] matches [Bill.id]
 * (`<type><number>-<congress>`, e.g. `hr1234-119`) and is null for
 * roll calls not tied to a bill (quorum calls, journal votes).
 */
@Serializable
data class RollCallVote(
    val id: String,
    val congress: Int,
    val chamber: Chamber,
    val session: Int,
    @SerialName("roll_number") val rollNumber: Int,
    val date: String,
    val question: String,
    val result: String,
    @SerialName("bill_id") val billId: String? = null,
    val totals: VoteTotals,
    val positions: List<MemberVote>,
    @SerialName("source_url") val sourceUrl: String? = null,
)

File: ios/Packages/ICCore/Sources/ICModels/RollCallVote.swift
public enum VoteChamber: String, Codable, Hashable, Sendable {
  case house
  case senate
  case unknown

  public init(from decoder: any Decoder) throws {
    let value = try decoder.singleValueContainer().decode(String.self)
    self = Self(rawValue: value) ?? .unknown
  }
}

public struct VoteTotals: Codable, Equatable, Hashable, Sendable {
  public let yea: Int
  public let nay: Int
  public let present: Int
  public let notVoting: Int

  public init(yea: Int, nay: Int, present: Int, notVoting: Int) {
    self.yea = yea
    self.nay = nay
    self.present = present
    self.notVoting = notVoting
  }

  private enum CodingKeys: String, CodingKey {
    case yea, nay, present
    case notVoting = "not_voting"
  }
}

public struct RollCallVoteReference: Codable, Equatable, Hashable, Identifiable, Sendable {
  public let id: String
  public let chamber: VoteChamber
  public let session: Int
  public let rollNumber: Int
  public let date: String
  public let question: String
  public let result: String
  public let billID: String?
  public let totals: VoteTotals
  public let partySplit: [String: [String: Int]]
  public let path: String

  public init(
    id: String,
    chamber: VoteChamber,
    session: Int,
    rollNumber: Int,
    date: String,
    question: String,
    result: String,
    billID: String? = nil,
    totals: VoteTotals,
    partySplit: [String: [String: Int]] = [:],
    path: String
  ) {
    self.id = id
    self.chamber = chamber
    self.session = session
    self.rollNumber = rollNumber
    self.date = date
    self.question = question
    self.result = result
    self.billID = billID
    self.totals = totals
    self.partySplit = partySplit
    self.path = path
  }

  private enum CodingKeys: String, CodingKey {
    case id, chamber, session, date, question, result, totals, path
    case rollNumber = "roll_number"
    case billID = "bill_id"
    case partySplit = "party_split"
  }

  public init(from decoder: any Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    id = try container.decode(String.self, forKey: .id)
    chamber = try container.decode(VoteChamber.self, forKey: .chamber)
    session = try container.decode(Int.self, forKey: .session)
    rollNumber = try container.decode(Int.self, forKey: .rollNumber)
    date = try container.decode(String.self, forKey: .date)
    question = try container.decode(String.self, forKey: .question)
    result = try container.decode(String.self, forKey: .result)
    billID = try container.decodeIfPresent(String.self, forKey: .billID)
    totals = try container.decode(VoteTotals.self, forKey: .totals)
    partySplit =
      try container.decodeIfPresent([String: [String: Int]].self, forKey: .partySplit) ?? [:]
    path = try container.decode(String.self, forKey: .path)
  }
}
